Hot Reload Performance Improvements

stuartwdouglas opened the pull request:
This PR will speed up the hot reload time for larger applications. Improvements likely won't really be noticeable in small demos, but it should have a big impact for larger applications, and lays the groundwork for further improvements.

- If no changes are made to the class structure then an attempt is made to replace the classes with the instrumentation API and avoid an application restart, resulting in pretty much instant replacement
- If changes are made the details of these changes are propagated into the build process, allowing information to be cached. This PR uses this info to cache Hibernate generated proxies and class based transformations, which can save a lot of time for larger applications.

gsmet commented:
This is 1.10 material, we need this to bake in master for a while.

Sanne commented on extensions/hibernate-orm/deployment/src/main/java/io/quarkus/hibernate/orm/deployment/HibernateOrmProcessor.java:
```diff
@@ -1194,37 +1211,44 @@ private PreGeneratedProxies generatedProxies(Set<String> entityClassNames, Index
-                    proxy = proxyAnnotations.get(subclassName);
+                CachedProxy result;
+                if (proxyCache.cache.containsKey(entity) && !changedClasses.contains(entity)) {
+                    result = proxyCache.cache.get(entity);
```
This looks very neat, but I believe you're not taking into account a change in related classes that would affect the output, such as `@Proxy(proxyClass -> changedInterface )` ?

Sanne replied:
And what about the current class inheriting further from other changed classes?

stuartwdouglas replied:
I now check the all classes and interfaces. If I check all interfaces then is that sufficient, or can @Proxy specify an interface the entity does not implement?

Sanne commented:
While I think it's a good idea to add a caching capability, let me know if the specific ORM enhancer has been seen as a problem. It's orthogonal, but I believe there's various things we could do to make it way leaner - I only didn't so far because of priorities, and ecause the things I have in mind are a bit tricky to get right. But less tricky than caching ;)

FroMage commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java:
```diff
@@ -409,6 +409,19 @@ public void handle(RoutingContext event) {
+                @Override
+                public void handle(HttpServerRequest event) {
+                    Thread.currentThread().setContextClassLoader(newCl);
+                    old.handle(event);
```
If `old.handle` was defined like this, it will still set its TCCL to the old TCCL it captured as the first thing it does right after you've just set the new TCCL, no?

stuartwdouglas replied:
Good point.

This is not actually supposed to be part of this PR, I have been working on this on and off for a while and I must have accidentally included an unrelated change. This has actually already been fixes as part of https://github.com/quarkusio/quarkus/commit/6b6db29a8f9af2146a497a688f359b78cabccabe

FroMage commented:
This is great!

I wonder if we couldn't do this a bit more declaratively, where build steps or class transformers / bytecode generators would declare what change they depend on.

To clarify, I wonder if the ORM build steps need to even run unless there's a change in any of its entity classes, or whatever config it requires.

So for instance, perhaps the ORM build steps could say:

- I depend on these config classes
- I depend on these classes that I've just scanned (which all have ORM annotations)
- I need to be called if any new class appears with these annotations

If it is true that the ORM build steps are so deterministic, then we could exclude perhaps find a way to not run them again and automatically cache whatever they produced last time?

This isn't as fine-grained as regenerating a single changed proxy class like what you've done here, but perhaps it could prevent large chunks of build steps from extensions unrelated to what the user changed, in a way that's relatively easy to implement for extension authors?

Sanne commented:
> I wonder if we couldn't do this a bit more declaratively, where build steps or class transformers / bytecode generators would declare what change they depend on.


I had exactly the same thought, but given the proxies we generate are rather "ad-hoc" to the ORM needs, I'm not sure how we could flesh this out. 

It might be true for class transformers, I wonder if there's a real need to have them opt-in with "cacheable" explicitly; could contrain them for pure transformations of input/output, making these unnecessary. I doubt we have any transformation who'd need to inject e.g. the current timestamp and I think we could be fine by documenting a limitation of expectations.

stuartwdouglas commented:
I think we might be able to build on this to make this declarative, and re-use recorded data, but IMHO this is a good first step.

In the profiling I have done the main dominator in terms of hot reload time for larger apps are generating bytecode and transformation, so I think there will likely be only a few places we need to apply this to get the low hanging fruit. If we want to take it further then we should look at a declarative approach and the ability to re-use recorded bytecode.

gsmet commented:
@Sanne could you have another look at this one? Thanks!

stuartwdouglas committed:
Add the option to perform hot reload via instrumentation

This does not require an app restart, so is much faster
for larger applications, but is limited to circumstances
where no structural changes have taken place.

This also adds the ability to check which classes have
been changed and what these changes actually are, which
allow frameworks to potentially cache metadata between
runs.

It also includes initial support for caching hibernate
proxies and transformations, as generating these are quite
expensive.

stuartwdouglas commented:
This has sat for over a month, and even though I opened it early in the 1.10 cycle to get feedback IMHO it is too late now. Can we please merge this ASAP so it actually makes 1.11? It provides a noticable user experience improvement for larger real world apps.

gsmet commented:
FYI, I pinged @Sanne again yesterday and he wanted to have another look, that's why I didn't merged it.

stuartwdouglas commented:
Can we get this merged? If @Sanne really wants changes they can be addressed later, but it has been waiting for his review for over a month.

FroMage commented:
This touches Hibernate so I understand why he wants to review it. However this only impacts Hibernate in DEV mode (in theory), so I think the impact is not dangerous. I've emailed Sanne just in case this slipped his radar. Let's see what happens.

Sanne commented:
Apologies for the delay. I remember having various concerns about this caching approach, but I don't feel I can complain without expressing them more precisely.

Honestly, I planned to create some tests to verify things but it's been a crazy time - I couldn't do it yet, and admittedly I'm not going to be able to do it today either.

So could someone else please review? As long as this only affects "dev mode" I'd say my concerns don't have to be blockers, I'll get back to it at better times.

geoand commented:
@Sanne what kind of tests did you have in mind?
